### PR TITLE
Fix Reader read/skip/mark implementation

### PR DESCRIPTION
**Pull request by exoego**
This PR fixes the Scala.js implementation of `Reader` classes so they behave in same manner as JRE implementation.

## 1. read
[`Reader#read()`](https://docs.oracle.com/javase/8/docs/api/java/io/Reader.html#read--) should return -1 if the end of the stream has been reached.

## 2. skip
Return value of [`Reader#skip(int)`](https://docs.oracle.com/javase/8/docs/api/java/io/Reader.html#skip-long-) is stated as `The number of characters actually skipped` so it suposed to be `0` after reaching the end.
This is also [documented in `StringReader`](https://docs.oracle.com/javase/8/docs/api/java/io/StringReader.html#skip-long-).

## 3. mark
[`StringReader#mark(int)`](https://docs.oracle.com/javase/8/docs/api/java/io/StringReader.html#mark-int-) and [`BufferedReader#mark(int)`](https://docs.oracle.com/javase/8/docs/api/java/io/BufferedReader.html#mark-int-) are supposed to throw `IllegalArgumentException` if received negative limit

## <del>4. constructor</del>
<del>Their constructors are not documented to throw NPE when receiving a null, but it should be as same as JRE implementation so that users can notice NPE on construction before using methods.</del>
In Scala.js's javalib emulation, we don't throw NPE explicitly https://github.com/scala-js/scala-js/pull/3665#issuecomment-495544361

### CODE REVIEW

**Comment by sjrd** [comment]
Please add an appropriate test in https://github.com/scala-js/scala-js/blob/0.6.x/test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala.

Since we are in the middle of a release, I'll look at this next week, probably.

**Comment by exoego** [comment]
Added tests and fixed `InputStreamReader` as well.

**Comment by sjrd** [comment]
> Their constructors are not documented to throw NPE when receiving a null, but it should be as same as JRE implementation so that users can notice NPE on construction before using methods.

NPEs are undefined behavior in Scala.js. The rule of thumb is not to explicitly check them. If they happen to throw a `TypeError`, fine. If they happen not to throw one, that's fine too. But we don't add code dedicated to throwing NPEs.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala [hunk 1]
```diff
@@ -293,4 +338,19 @@ class InputStreamReaderTest {
     assertEquals(0, streamReader.read(new Array[Char](0)))
   }
 
+  @Test def should_always_return_0_after_reaching_end(): Unit = {
+    val data = "Lorem ipsum".getBytes()
+    val r = new InputStreamReader(new ByteArrayInputStream(data))
+    assertEquals(1, r.skip(100))
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala]: This line fails on the JVM (`testSuiteJVM/test`).

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala [hunk 2]
```diff
@@ -293,4 +338,19 @@ class InputStreamReaderTest {
...
+    assertEquals(-1, r.read())
+
+    assertEquals(0, r.skip(100) )
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala]: Spurious space.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala [hunk 3]
```diff
@@ -112,13 +115,32 @@ class StringReaderTest {
...
     assertTrue(newReader.markSupported)
   }
+
+  @Test def should_always_return_0_after_reaching_end(): Unit = {
+    val r = newReader
+    assertEquals(4, r.skip(100))
+    assertEquals(-1, r.read())
+
+    // StringReader accepts negative skip
+    assertEquals(0, r.skip(-100))
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala]: Could you also add a test where the negative skip does not try to go back more than the full available length in the string? I.e., with an absolute value that is less than the string length.

For example `-2` in this case?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala [hunk 4]
```diff
@@ -13,19 +13,22 @@
 package org.scalajs.testsuite.javalib.io
 
 import scala.annotation.tailrec
-
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala]: This blank line should not disappear.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala [hunk 5]
```diff
@@ -13,19 +13,22 @@
...
 import java.io._
+import java.nio.charset.Charset
 
 import org.junit.Test
 import org.junit.Assert._
-
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala]: This blank line should not disappear.

File: javalib/src/main/scala/java/io/StringReader.scala [hunk 6]
```diff
@@ -23,6 +27,9 @@ class StringReader(s: String) extends Reader {
   }
 
   override def mark(readAheadLimit: Int): Unit = {
+    if (readAheadLimit < 0) {
+      throw new IllegalArgumentException("Read-ahead limit < 0")
+    }
```
Comment (sjrd) [inline javalib/src/main/scala/java/io/StringReader.scala]: No need for braces since there is only one statement in the `if` branch. See https://github.com/scala-js/scala-js/blob/master/CODINGSTYLE.md#braces

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala [hunk 7]
```diff
@@ -245,10 +267,33 @@ class BufferedReaderTest {
   @Test def should_support_marking(): Unit = {
     assertTrue(newReader.markSupported)
   }
+
+  @Test def should_always_return_0_after_reaching_end(): Unit = {
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala]:
```suggestion
  @Test def skip_should_always_return_0_after_reaching_end(): Unit = {
```

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala [hunk 8]
```diff
@@ -112,13 +115,32 @@ class StringReaderTest {
...
+    val r = newReader
+    assertEquals(4, r.skip(100))
+    assertEquals(-1, r.read())
+
+    // StringReader accepts negative skip
+    assertEquals(0, r.skip(-100))
+    assertEquals(-1, r.read())
+  }
+
+  @Test def should_throw_IOEx(): Unit = {
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala]:
```suggestion
  @Test def mark_should_throw_with_negative_lookahead(): Unit = {
```

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala [hunk 9]
```diff
@@ -112,13 +115,32 @@ class StringReaderTest {
   @Test def should_support_marking(): Unit = {
     assertTrue(newReader.markSupported)
   }
+
+  @Test def should_always_return_0_after_reaching_end(): Unit = {
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala]:
```suggestion
  @Test def skip_should_always_return_0_after_reaching_end(): Unit = {
```

**Comment by exoego** [comment]
> 
> 
> > Their constructors are not documented to throw NPE when receiving a null, but it should be as same as JRE implementation so that users can notice NPE on construction before using methods.
> 
> NPEs are undefined behavior in Scala.js. The rule of thumb is not to explicitly check them. If they happen to throw a `TypeError`, fine. If they happen not to throw one, that's fine too. But we don't add code dedicated to throwing NPEs.

OK.
I skipped the commit for NPE

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala (hunk 3 above)
Comment (exoego) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala]: Added a test dedicated for negative lookahead in https://github.com/scala-js/scala-js/pull/3665/commits/570c9f3c3a842224a26c5d1fcf8c3f9e0890e881#diff-6172b8f08f1ddaf5a345c353dbec80b0R116

File: javalib/src/main/scala/java/io/InputStreamReader.scala [hunk 10]
```diff
@@ -72,6 +72,7 @@ class InputStreamReader(private[this] var in: InputStream,
     ensureOpen()
 
     if (outBuf.hasRemaining) outBuf.get()
+    else if (in.available() <= 0) -1
```
Comment (exoego) [inline javalib/src/main/scala/java/io/InputStreamReader.scala]: [`ZipInputStreamTest` fails due to this change](https://scala-webapps.epfl.ch/sjsjenkins/blue/organizations/jenkins/scalajs-main/detail/PR-3665/8/pipeline#step-96-log-111).
>[error] Test scala.scalajs.testsuite.javalibex.ZipInputStreamTest.should_read_zip_archives failed: java.lang.AssertionError: expected:<101> but was:<65535>, took 0.010000128 sec
> [error]     at scala.scalajs.testsuite.javalibex.ZipInputStreamTest.expectStrEntry$1(/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-3665/javalib-ex-test-suite/target/scala-2.10/file:/localhome/jenkins/sjsjenkins/workspace/scalajs-main_PR-3665/javalib-ex-test-suite/src/test/scala/scala/scalajs/testsuite/javalibex/ZipInputStreamTest.scala:48)

It could be potential issue in `ZipInputStream`.
I will investigate.

Comment (exoego) [reply javalib/src/main/scala/java/io/InputStreamReader.scala]: I think this was due to inappropriate logic of `Reader.skip` and is fixed in https://github.com/scala-js/scala-js/pull/3665/commits/e965d0259653874e150bfebe1bed781850e5c338

Comment (exoego) [reply javalib/src/main/scala/java/io/InputStreamReader.scala]: I think this was due to inappropriate implementation of `Reader#skip` that is inherited by `InputStreamReader`.
Fixed in https://github.com/scala-js/scala-js/pull/3665/commits/e965d0259653874e150bfebe1bed781850e5c338

File: javalib/src/main/scala/java/io/BufferedReader.scala [hunk 11]
```diff
@@ -36,6 +36,7 @@ class BufferedReader(in: Reader, sz: Int) extends Reader {
   }
 
   override def mark(readAheadLimit: Int): Unit = {
+    if (readAheadLimit < 0) throw new IllegalArgumentException("Read-ahead limit < 0")
```
Comment (sjrd) [inline javalib/src/main/scala/java/io/BufferedReader.scala]: The `throw` should be on a new line, indented 2 spaces, because the `if` is in statement position. See https://github.com/scala-js/scala-js/blob/master/CODINGSTYLE.md#ifelse.

File: javalib/src/main/scala/java/io/StringReader.scala [hunk 12]
```diff
@@ -23,6 +23,7 @@ class StringReader(s: String) extends Reader {
   }
 
   override def mark(readAheadLimit: Int): Unit = {
+    if (readAheadLimit < 0) throw new IllegalArgumentException("Read-ahead limit < 0")
```
Comment (sjrd) [inline javalib/src/main/scala/java/io/StringReader.scala]: The `throw` should be on a new line, indented 2 spaces, because the `if` is in statement position. See https://github.com/scala-js/scala-js/blob/master/CODINGSTYLE.md#ifelse.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala [hunk 13]
```diff
@@ -112,6 +112,32 @@ class StringReaderTest {
   @Test def should_support_marking(): Unit = {
     assertTrue(newReader.markSupported)
   }
+
+  @Test def skip_should_accept_negative_lookahead_as_lookback(): Unit = {
+    // StringReader.skip accepts negative lookahead
+    val r = newReader
+    assertEquals("already head",0, r.skip(-1))
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala]: Missing space after comma.

Comment (exoego) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala]: Fixed.

File: javalib/src/main/scala/java/io/StringReader.scala (hunk 12 above)
Comment (exoego) [reply javalib/src/main/scala/java/io/StringReader.scala]: Fixed

File: javalib/src/main/scala/java/io/BufferedReader.scala (hunk 11 above)
Comment (exoego) [reply javalib/src/main/scala/java/io/BufferedReader.scala]: Fixed

File: javalib/src/main/scala/java/io/Reader.scala [hunk 14]
```diff
@@ -53,7 +53,7 @@ abstract class Reader private[this] (_lock: Option[Object])
   def skip(n: Long): Long = {
     if (n < 0)
       throw new IllegalArgumentException("Cannot skip negative amount")
-    else if (read() == -1) 0
+    else if (read(Array.ofDim(n.toInt), 0, n.toInt) == -1) 0
     else 1
```
[the remark this answers is not in the document]
Comment (exoego) [reply javalib/src/main/scala/java/io/Reader.scala]: Added loop to support `n > Int.MaxValue` in https://github.com/scala-js/scala-js/pull/3665/commits/5becb3f424334331c7ed47dd8e2a070b0eb712f7

File: javalib/src/main/scala/java/io/Reader.scala [hunk 15]
```diff
@@ -53,8 +55,20 @@ abstract class Reader private[this] (_lock: Option[Object])
...
     if (n < 0)
       throw new IllegalArgumentException("Cannot skip negative amount")
-    else if (read() == -1) 0
-    else 1
+    else {
+      @tailrec
+      def loop(m: Long, lastSkipped: Long): Long = {
+        if (m <= 0)
+          lastSkipped
+        else {
```
Comment (sjrd) [inline javalib/src/main/scala/java/io/Reader.scala]: Same thing with the braces.

File: javalib/src/main/scala/java/io/Reader.scala [hunk 16]
```diff
@@ -53,8 +55,20 @@ abstract class Reader private[this] (_lock: Option[Object])
...
+          val mMin = Math.min(m, 8192).toInt
+          val skipped = Math.max(read(new Array[Char](mMin), 0, mMin), 0)
```
Comment (sjrd) [inline javalib/src/main/scala/java/io/Reader.scala]: Consider allocating this array only once, outside of the loop, for example with
```scala
val buffer = new Array[Char](Math.min(n, 8192).toInt)
```

Comment (sjrd) [reply javalib/src/main/scala/java/io/Reader.scala]: Also, I think the new implementation fails to stop the loop if the input runs into EOF before `n` chars have been skipped. You'll need to stop the loop if `read` returns `< 0` (and return `lastSkipped` in that case).

(gosh, looks like this method is much trickier than it looks at first)

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala [hunk 17]
```diff
@@ -22,6 +22,21 @@ import org.junit.Assert._
...
+    def read(dbuf: Array[Char], off: Int, len: Int): Int = {
+      java.util.Arrays.fill(dbuf, off, off + len, 'A')
+      len
+    }
+    def close(): Unit = ()
+  }
+
+  @Test def skip_should_always_skip_n_if_possible(): Unit = {
+    assertEquals(42, MyReader.skip(42))
+    assertEquals(2147483700L, MyReader.skip(Int.MaxValue.toLong + 53))
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala]: Apparently, this test takes way too much time on Rhino, which is why the CI failed. Please add the following at the beginning of the method to fix that:
```scala
assumeFalse("Too slow in Rhino", executingInRhino)
```
with the following imports at the top:
```scala
import org.junit.Assume._

import org.scalajs.testsuite.utils.Platform._
```

Comment (exoego) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala]: Applied the changes in https://github.com/scala-js/scala-js/pull/3665/commits/d51ba432689c13b324e1914f00770bfa4aeea813

Comment (sjrd) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala]: Well, apparently this test is too slow, period. Not just with Rhino. I tried running the CI a few times on this PR, but it keeps failing.

So could you add one test with a skip greater than 8192 but still small (e.g., 10000) so that there is a test that exercises the loop, but remove the test with 2147483700L altogether (and then also remove the assumption and the corresponding imports)?

Thanks.

Comment (sjrd) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/ReadersTest.scala]: I did that small change and force-pushed, so that I can merge this PR before going on vacation.
